File: control/api/src/routes/feedback.ts
// POST /api/feedback - Submit feedback
// POST /api/feedback/summarize - Trigger feedback summarization

import { Router } from 'express';
import { fetchFromPython, checkPythonHealth } from '../services/python-client.js';
import type { FeedbackRequest, FeedbackResponse, SummarizeRequest, SummarizeResponse } from '../../../shared/schema/api-contracts.js';

const router = Router();

router.post('/feedback', async (req, res) => {
  try {
    const { sessionId, styleId, feedback } = req.body as FeedbackRequest;

    if (!sessionId || !styleId || !feedback) {
      return res.status(400).json({
        success: false,
        error: {
          code: 'INVALID_REQUEST',
          message: 'Missing required fields: sessionId, styleId, and feedback are required'
        }
      });
    }

    const pythonHealthy = await checkPythonHealth();
    if (!pythonHealthy) {
      return res.status(503).json({
        success: false,
        error: {
          code: 'BACKEND_UNAVAILABLE',
          message: 'Python backend is unavailable'
        }
      });
    }

    const response = await fetchFromPython<FeedbackResponse>('/feedback', {
      method: 'POST',
      body: JSON.stringify({ sessionId, styleId, feedback })
    });

    return res.json(response);
  } catch (error) {
    console.error('Feedback error:', error);
    res.status(500).json({
      success: false,
      error: {
        code: 'FEEDBACK_ERROR',
        message: error instanceof Error ? error.message : 'Unknown error'
      }
    });
  }
});

router.post('/feedback/summarize', async (req, res) => {
  try {
    const { sessionId, styleId } = req.body as SummarizeRequest;

    if (!sessionId || !styleId) {
      return res.status(400).json({
        success: false,
        error: {
          code: 'INVALID_REQUEST',
          message: 'Missing required fields: sessionId and styleId are required'
        }
      });
    }

    const pythonHealthy = await checkPythonHealth();
    if (!pythonHealthy) {
      return res.status(503).json({
        success: false,
        error: {
          code: 'BACKEND_UNAVAILABLE',
          message: 'Python backend is unavailable'
        }
      });
    }

    const response = await fetchFromPython<SummarizeResponse>('/feedback/summarize', {
      method: 'POST',
      body: JSON.stringify({ sessionId, styleId })
    });

    return res.json(response);
  } catch (error) {
    console.error('Summarize error:', error);
    res.status(500).json({
      success: false,
      error: {
        code: 'SUMMARIZE_ERROR',
        message: error instanceof Error ? error.message : 'Unknown error'
      }
    });
  }
});

export default router;

File: api/routes/feedback.py
"""Feedback endpoints for conversation context and summarization."""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from api.services.pipeline import PipelineService

router = APIRouter()


class FeedbackRequest(BaseModel):
    """Request body for submitting feedback."""
    sessionId: str
    styleId: str
    feedback: str


class SummarizeRequest(BaseModel):
    """Request body for triggering feedback summarization."""
    sessionId: str
    styleId: str


@router.post("/feedback")
def submit_feedback(request: FeedbackRequest):
    """Submit designer feedback for the current session."""
    try:
        if not request.feedback or not request.feedback.strip():
            raise HTTPException(status_code=400, detail="Feedback cannot be empty")

        service = PipelineService()
        turn_number, was_summarized = service.add_feedback(
            session_id=request.sessionId,
            style_id=request.styleId,
            feedback=request.feedback,
        )

        return {
            "success": True,
            "turnNumber": turn_number,
            "summarized": was_summarized
        }

    except HTTPException:
        raise
    except Exception as e:
        return {
            "success": False,
            "error": {
                "code": "FEEDBACK_ERROR",
                "message": str(e)
            }
        }


@router.post("/feedback/summarize")
def summarize_feedback(request: SummarizeRequest):
    """Trigger GPT summarization of accumulated feedback and persist to style."""
    try:
        service = PipelineService()
        summary = service.summarize_feedback(
            session_id=request.sessionId,
            style_id=request.styleId
        )

        return {
            "success": True,
            "summary": summary
        }

    except Exception as e:
        return {
            "success": False,
            "error": {
                "code": "SUMMARIZE_ERROR",
                "message": str(e)
            }
        }
